tests/utils/component_test_utils.go

```diff
@@ -0,0 +1,233 @@
+		} else if testComponent.ComponentType == schema.VolumeComponentType {
+			setVolumeComponentValues(component.Volume)
+		}
+	} else {
+		errorString = append(errorString, LogInfoMessage(fmt.Sprintf("....... Component not found in test : %s", component.Name)))
+	}
+	var err error
+	if len(errorString) > 0 {
+		err = errors.New(fmt.Sprint(errorString))
+	}
+	return err
+}
+
+// VerifyComponents verifies components returned by the parser are the same as those saved in the devfile schema
+func (devfile TestDevfile) VerifyComponents(parserComponents []schema.Component) error {
+
+	LogInfoMessage("Enter VerifyComponents")
+	var errorString []string
+
+	// Compare entire array of components
+	if !cmp.Equal(parserComponents, devfile.SchemaDevFile.Components) {
+		errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component array compare failed.")))
+		for _, component := range parserComponents {
+			if testComponent, found := getSchemaComponent(devfile.SchemaDevFile.Components, component.Name); found {
+				if !cmp.Equal(component, *testComponent) {
+					parserFilename := AddSuffixToFileName(devfile.FileName, "_"+component.Name+"_Parser")
+					testFilename := AddSuffixToFileName(devfile.FileName, "_"+component.Name+"_Test")
+					LogInfoMessage(fmt.Sprintf(".......marshall and write devfile %s", parserFilename))
+					c, err := yaml.Marshal(component)
+					if err != nil {
+						errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......marshall devfile %s", parserFilename)))
+					} else {
+						err = ioutil.WriteFile(parserFilename, c, 0644)
+						if err != nil {
+							errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......write devfile %s", parserFilename)))
+						}
+					}
+					LogInfoMessage(fmt.Sprintf(".......marshall and write devfile %s", testFilename))
+					c, err = yaml.Marshal(testComponent)
+					if err != nil {
+						errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......marshall devfile %s", testFilename)))
+					} else {
+						err = ioutil.WriteFile(testFilename, c, 0644)
+						if err != nil {
+							errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......write devfile %s", testFilename)))
+						}
+					}
+					errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component %s did not match, see files : %s and %s", component.Name, parserFilename, testFilename)))
+				} else {
+					LogInfoMessage(fmt.Sprintf(" --> Component matched : %s", component.Name))
+				}
+			} else {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component from parser not known to test - id : %s ", component.Name)))
+			}
+		}
+		for _, component := range devfile.SchemaDevFile.Components {
+			if _, found := getSchemaComponent(parserComponents, component.Name); !found {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component from test not returned by parser : %s ", component.Name)))
+			}
+		}
```

Comment:
Same as the command verify comment.

Reply:
The previous loop goes through all the components returned by the parser and makes sure they are the same as those expected by test. This loop does a quick check that the test does not have components that the parser did not return.